File: Ruby/Euler012.rb
#!/usr/bin/env ruby

def main
  target = 500
  triangle = 0
  incrementer = 1
  while true
    triangle += incrementer
    if get_factors(triangle).length >= target
      puts triangle
      break
    end
    incrementer += 1
  end
end

def get_factors(of_int)
  max = Math.sqrt(of_int).ceil

  factors = (1..max).select { |factor| of_int % factor == 0 }
      .map { |factor| [factor, of_int / factor] }
      .flatten
      .uniq

  return factors
end

if __FILE__ == $0
  main
end

File: Node/Euler012.js
#!/usr/bin/env node
const generators = require("./generators.js");

function main() {
  const target = 500;

  var triangles = generators.filter(generate_triangles(), (t) => { return getFactors(t).size > target });
  for (let t of triangles) {
    console.log(t);
    break;
  }
}

function getFactors(ofInt) {
  const max = Math.ceil(Math.sqrt(ofInt));
  var factors = new Set();
  for (var factor = 1; factor <= max; factor++) {
    if (ofInt % factor == 0) {
      factors.add(factor);
      factors.add(ofInt / factor);
    }
  }
  return factors;
}

function* generate_triangles() {
  var triangle = 0;
  var incrementer = 1;
  while (true) {
    triangle += incrementer;
    yield triangle;
    incrementer++;
  }
}

if (require.main == module) {
  main();
}
